File: leetcode/Python/0133-m-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, neighbors):
        self.val = val
        self.neighbors = neighbors
"""

"""
BFS
"""
class Solution_1:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return
        nodeCopy = Node(node.val, [])
        dic = {node: nodeCopy}
        self.bfs(node, dic)
        return nodeCopy
    
    def bfs(self, node, dic):
        q = collections.deque([node])
        while q:
            node = q.pop()
            for neighbor in node.neighbors:
                if neighbor not in dic:
                    neighborCopy = Node(neighbor.val, [])
                    dic[neighbor] = neighborCopy
                    dic[node].neighbors.append(neighborCopy)
                    q.append(neighbor)
                else:
                    dic[node].neighbors.append(dic[neighbor])
"""
DFS
"""
class Solution_2:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return 
        nodeCopy = Node(node.val, [])
        dic = {node: nodeCopy}
        self.dfs(node, dic)
        return nodeCopy

    def dfs(self, node, dic):
        for neighbor in node.neighbors:
            if neighbor not in dic:
                neighborCopy = Node(neighbor.val, [])
                dic[neighbor] = neighborCopy
                dic[node].neighbors.append(neighborCopy)
                self.dfs(neighbor, dic)
            else:
                dic[node].neighbors.append(dic[neighbor])

File: leetcode/C++/0133-m-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node)
            return NULL;
        Node* nodeCopy = new Node(node->val);
        unordered_map<Node*, Node*> mp;
        mp[node] = nodeCopy;
        
        queue<Node*> q;
        q.push(node);
        while(!q.empty()) {
            Node* nd = q.front();
            q.pop();
            for (Node* nb : nd->neighbors) {
                if (mp.find(nb) == mp.end()) {
                    Node* nbCopy = new Node(nb->val);
                    mp[nb] = nbCopy;
                    mp[nd]->neighbors.push_back(nbCopy);
                    q.push(nb);
                } else {
                    mp[nd]->neighbors.push_back(mp[nb]);
                }
            }
        }
        return nodeCopy;
    }
};